### PR TITLE
Replicates psispace in rust

### PR DESCRIPTION
**Pull request by sparkplug0025**
- Replicates psispace-sys in rust
- Implements table.erase
- Adds number of secondary keys in the `Table` trait (to optimize erase operation)
- General improvements on the election service (now it supports current time and serves the simple ui)
- Improves table interface adding the methods: `new(), with_service() and with_prefix()`

### CODE REVIEW

File: rust/test_service_elections/src/lib.rs [hunk 1]
```diff
@@ -71,6 +71,10 @@ mod service {
         }
     }
 
+    fn get_current_time() -> TimePointSec {
+        transaction_sys::Wrapper::call().headBlockTime()
```
Comment (tbfleming) [inline rust/test_service_elections/src/lib.rs:75]: headBlockTime is marked `TODO: remove`. Use `currentBlock()`, which is more suitable than `headBlock()` for most cases.

File: rust/psibase/src/table.rs [hunk 2]
```diff
@@ -46,25 +47,24 @@ pub trait TableRecord: PackableOwned {
...
 
-    fn open() -> Self {
-        let prefix = (Self::TABLE_SERVICE, Self::TABLE_INDEX).to_key();
+    fn new() -> Self {
+        let prefix = (get_service(), Self::TABLE_INDEX).to_key();
         Self::create_table_from_prefix(prefix)
     }
 
-    fn open_custom(service: AccountNumber, table_index: u16) -> Self {
+    fn with_service(service: AccountNumber, table_index: u16) -> Self {
```
Comment (tbfleming) [inline rust/psibase/src/table.rs:61]: remove `table_index`. This function handles any time service A reads service B's tables.

File: rust/psibase/src/table.rs [hunk 3]
```diff
@@ -46,25 +47,24 @@ pub trait TableRecord: PackableOwned {
...
         let prefix = (service, table_index).to_key();
         Self::create_table_from_prefix(prefix)
     }
 
     fn create_table_from_prefix(prefix: Vec<u8>) -> Self {
```
Comment (tbfleming) [inline rust/psibase/src/table.rs:66]: Users providing the prefix is a rare case, so having just the 2-argument version is enough.

**Review by tbfleming — APPROVED** [review]

File: rust/psibase/src/table.rs (hunk 2 above)
Comment (sparkplug0025) [reply rust/psibase/src/table.rs:61]: What if service A, for convenience, also have the same table as table B under a different index? Eg. I could see services reusing the definition from `psispace-sys.ContentTable`. Should we always encourage duplicating the declaration?

File: rust/psibase/src/table.rs (hunk 3 above)
Comment (sparkplug0025) [reply rust/psibase/src/table.rs:66]: the `create_table_from_prefix` is used from the `with_service()` and also from `new()`. I did not mean to expose this method but I guess we can't hide it in a trait.

**Comment by sparkplug0025** [comment]
I'm merging but I can address above comments in a next PR.

File: rust/psibase/src/table.rs (hunk 2 above)
Comment (tbfleming) [reply rust/psibase/src/table.rs:61]: See https://doc-sys.psibase.io/rust-service/tables.html#storing-structs-defined-elsewhere

File: rust/psibase/src/table.rs (hunk 3 above)
Comment (tbfleming) [reply rust/psibase/src/table.rs:66]: It's small enough to inline in the other functions

File: rust/psibase/src/table.rs (hunk 2 above)
Comment (sparkplug0025) [reply rust/psibase/src/table.rs:61]: I was working in the Global WebContent table and I realized that the consumer does not need to redeclare a shared struct keys like in the docs example.

Here's the WebContent global record struct: 
https://github.com/gofractally/psibase/pull/222/files#diff-64aad0707509e772a7c54f564cd388b06db283a4d6bb5042e9bcdd548a5d96cdR219

And here's how a simplified macro could be (to replace the following lines Table impl):
https://github.com/gofractally/psibase/pull/222/files#diff-ed73e7a1a9fad9c7b0753d5de5101331739bff714eef72fd1a054cbe72ba8e4cR75-R77

What do you think? Should I pursue this macro?

**Pull request merged**